File: client/android/app/src/main/java/com/clawdbot/android/mchat/MChatConnection.kt
package com.clawdbot.android.mchat

import android.util.Log
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import org.eclipse.paho.client.mqttv3.IMqttActionListener
import org.eclipse.paho.client.mqttv3.IMqttToken
import org.eclipse.paho.client.mqttv3.MqttAsyncClient
import org.eclipse.paho.client.mqttv3.MqttCallback
import org.eclipse.paho.client.mqttv3.MqttConnectOptions
import org.eclipse.paho.client.mqttv3.MqttException
import org.eclipse.paho.client.mqttv3.MqttMessage
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

private const val TAG = "MChatConnection"
private const val REQ_PREFIX = "mchat/msg/req/"
private const val RESP_PREFIX = "mchat/msg/resp/"
private const val INBOX_PREFIX = "mchat/inbox/"
private const val QOS = 1
private val UTF8 = Charsets.UTF_8
private const val REQUEST_TIMEOUT_MS = 30_000L

sealed class MChatConnectionState {
  data object Disconnected : MChatConnectionState()
  data object Connecting : MChatConnectionState()
  data object Connected : MChatConnectionState()
  data class Error(val message: String) : MChatConnectionState()
}

/**
 * MChat MQTT 连接：使用员工连接信息连接 Broker，auth.bind，订阅收件箱，支持 request(action, params) 与收件箱消息回调。
 */
class MChatConnection(
  private val scope: CoroutineScope,
  brokerHost: String,
  brokerPort: Int,
  useTls: Boolean,
  username: String,
  password: String,
  employeeId: String,
  clientId: String? = null,
) {
  @Volatile private var onInboxMessage: ((String) -> Unit)? = null
  fun setInboxCallback(callback: (String) -> Unit) { onInboxMessage = callback }
  private val connectUsername = username
  private val connectPassword = password
  private val clientIdValue = clientId?.trim()?.takeIf { it.isNotEmpty() }
    ?: "${employeeId}_android_${UUID.randomUUID().toString().take(8)}"

  private val brokerUrl = buildString {
    append(if (useTls) "ssl://" else "tcp://")
    append(brokerHost.trim())
    append(":")
    append(brokerPort.coerceIn(1, 65535))
  }

  private val _state = MutableStateFlow<MChatConnectionState>(MChatConnectionState.Disconnected)
  val state: StateFlow<MChatConnectionState> = _state.asStateFlow()

  private val pending = ConcurrentHashMap<String, CompletableDeferred<MqttResponse>>()
  @Volatile private var client: MqttAsyncClient? = null
  @Volatile private var connectionDeferred: CompletableDeferred<Unit>? = null

  private val employeeIdValue = employeeId.trim()
  private val respTopicPrefix = "$RESP_PREFIX$clientIdValue/"
  private val inboxTopic = "$INBOX_PREFIX$employeeIdValue"

  data class MqttResponse(val code: Int, val message: String, val data: String?)

  fun connect() {
    scope.launch(Dispatchers.IO) {
      doConnect(connectUsername, connectPassword)
    }
  }

  fun disconnect() {
    scope.launch(Dispatchers.IO) {
      try {
        client?.disconnect()?.waitForCompletion(2000)
        client?.close(true)
      } catch (_: Throwable) {}
      client = null
      connectionDeferred = null
      pending.values.forEach { it.completeExceptionally(IllegalStateException("Disconnected")) }
      pending.clear()
      _state.value = MChatConnectionState.Disconnected
    }
  }

  suspend fun request(action: String, params: Map<String, Any?>): MqttResponse {
    ensureConnected()
    val seqId = "seq_${UUID.randomUUID().toString().take(8)}_${System.currentTimeMillis()}"
    val topic = "${REQ_PREFIX}$clientIdValue/$seqId"
    val payload = buildString {
      append("{\"action\":\"$action\"")
      params.forEach { (k, v) ->
        when (v) {
          null -> append(",\"$k\":null")
          is String -> append(",\"$k\":${jsonEscape(v)}")
          is Number -> append(",\"$k\":$v")
          is Boolean -> append(",\"$k\":$v")
          else -> append(",\"$k\":\"$v\"")
        }
      }
      append("}")
    }
    val deferred = CompletableDeferred<MqttResponse>()
    pending[seqId] = deferred
    try {
      val c = client ?: throw IllegalStateException("Not connected")
      if (!c.isConnected) throw IllegalStateException("Not connected")
      c.publish(topic, payload.toByteArray(UTF8), QOS, false)
      return withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
        deferred.await()
      } ?: MqttResponse(504, "Request timeout", null)
    } finally {
      pending.remove(seqId)
    }
  }

  private suspend fun ensureConnected() {
    val c = client
    if (c?.isConnected == true) return
    val def = connectionDeferred ?: throw IllegalStateException("Not connected")
    def.await()
  }

  private suspend fun doConnect(username: String, password: String) = withContext(Dispatchers.IO) {
    if (client?.isConnected == true) return@withContext
    _state.value = MChatConnectionState.Connecting
    val deferred = CompletableDeferred<Unit>()
    connectionDeferred = deferred

    val opts = MqttConnectOptions().apply {
      isCleanSession = true
      keepAliveInterval = 60
      connectionTimeout = 30
      mqttVersion = MqttConnectOptions.MQTT_VERSION_3_1_1
      this.userName = username
      this.password = password.toCharArray()
    }

    val mqtt = MqttAsyncClient(brokerUrl, clientIdValue, MemoryPersistence())
    client = mqtt

    mqtt.setCallback(object : MqttCallback {
      override fun connectionLost(cause: Throwable?) {
        Log.w(TAG, "connectionLost: ${cause?.message}")
        client = null
        connectionDeferred = null
        pending.values.forEach { it.cancel() }
        pending.clear()
        _state.value = MChatConnectionState.Disconnected
      }

      override fun messageArrived(topic: String, message: MqttMessage) {
        val raw = message.payload?.toString(UTF8) ?: return
        val payload = raw.replace("\u0000", "").trim()
        if (payload.isEmpty()) return

        when {
          topic.startsWith(respTopicPrefix) -> {
            val seqId = topic.removePrefix(respTopicPrefix)
            val deferred = pending[seqId] ?: return
            try {
              val code = parseCode(payload)
              val msg = parseMessage(payload)
              val data = parseData(payload)
              deferred.complete(MqttResponse(code, msg, data))
            } catch (e: Throwable) {
              deferred.complete(MqttResponse(500, e.message ?: "Parse error", null))
            }
          }
          topic == inboxTopic -> onInboxMessage?.invoke(payload)
        }
      }

      override fun deliveryComplete(token: org.eclipse.paho.client.mqttv3.IMqttDeliveryToken?) {}
    })

    try {
      mqtt.connect(opts).waitForCompletion(15_000)
      if (!mqtt.isConnected) {
        _state.value = MChatConnectionState.Error("Connect failed")
        deferred.completeExceptionally(IllegalStateException("Not connected"))
        return@withContext
      }
      mqtt.subscribe("$respTopicPrefix+", QOS).waitForCompletion(5_000)
      mqtt.subscribe(inboxTopic, QOS).waitForCompletion(5_000)

      val bindRes = request("auth.bind", mapOf("employee_id" to employeeIdValue))
      if (bindRes.code != 0) {
        _state.value = MChatConnectionState.Error("auth.bind: ${bindRes.message}")
        deferred.completeExceptionally(IllegalStateException(bindRes.message))
        return@withContext
      }

      _state.value = MChatConnectionState.Connected
      deferred.complete(Unit)
    } catch (e: Throwable) {
      Log.e(TAG, "connect failed", e)
      _state.value = MChatConnectionState.Error(e.message ?: "Connect failed")
      client = null
      connectionDeferred = null
      deferred.completeExceptionally(e)
    }
  }

  private fun parseCode(payload: String): Int {
    val i = payload.indexOf("\"code\"")
    if (i < 0) return 500
    val start = payload.indexOf(':', i)
    if (start < 0) return 500
    val end = payload.indexOf(',', start).let { if (it < 0) payload.indexOf('}', start) else it }
    return payload.substring(start + 1, end).trim().toIntOrNull() ?: 500
  }

  private fun parseMessage(payload: String): String {
    val i = payload.indexOf("\"message\"")
    if (i < 0) return ""
    val start = payload.indexOf('"', i + 10)
    if (start < 0) return ""
    val end = payload.indexOf('"', start + 1)
    return if (end > start) payload.substring(start + 1, end) else ""
  }

  private fun parseData(payload: String): String? {
    val i = payload.indexOf("\"data\"")
    if (i < 0) return null
    val colon = payload.indexOf(':', i)
    if (colon < 0) return null
    val rest = payload.substring(colon + 1).trim()
    return when {
      rest.startsWith("null") -> null
      rest.startsWith("{") || rest.startsWith("[") -> {
        var depth = 0
        var j = 0
        for (c in rest) {
          when (c) {
            '{', '[' -> depth++
            '}', ']' -> { depth--; if (depth == 0) return rest.substring(0, j + 1) }
          }
          j++
        }
        null
      }
      else -> null
    }
  }

  private fun jsonEscape(s: String): String =
    "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n") + "\""
}

File: client/ios/MoltChat/MChat/MChatConnection.swift
import Foundation
import Combine
import CocoaMQTT

/// MChat MQTT 连接，对标 Android MChatConnection
enum MChatConnectionState: Equatable {
    case disconnected
    case connecting
    case connected
    case error(String)
}

struct MqttResponse {
    let code: Int
    let message: String
    let data: String?
}

/// 根据 serviceId 生成 topic 前缀：有 serviceId 则 "{serviceId}/mchat"，否则 "mchat"
private func getTopicPrefix(_ serviceId: String?) -> String {
    guard let sid = serviceId?.trimmingCharacters(in: .whitespacesAndNewlines), !sid.isEmpty else {
        return "mchat"
    }
    return "\(sid)/mchat"
}

@MainActor
final class MChatConnection {
    private let brokerHost: String
    private let brokerPort: UInt16
    private let useTls: Bool
    private let username: String
    private let password: String
    private let employeeId: String
    /// 当前连接使用的员工 ID，用于按身份隔离历史会话缓存。
    var myEmployeeId: String { employeeId }
    private let clientIdValue: String
    /// 服务实例 ID，用于 Topic 域隔离
    private let serviceIdValue: String?
    
    private let topicPrefix: String
    private let reqPrefix: String
    private let respPrefixBase: String
    private let inboxPrefixBase: String
    private let requestTimeout: TimeInterval = 30
    private let initialReconnect: TimeInterval = 2
    private let maxReconnect: TimeInterval = 30
    
    private var mqtt: CocoaMQTT?
    private var onInboxMessage: ((String) -> Void)?
    private var pending: [String: CheckedContinuation<MqttResponse, Never>] = [:]
    private var reconnectTask: Task<Void, Never>?
    private var reconnectBackoff: TimeInterval = 2
    private var userDisconnect = false
    
    private(set) var state: MChatConnectionState = .disconnected {
        didSet { stateSubject.send(state) }
    }
    let stateSubject = CurrentValueSubject<MChatConnectionState, Never>(.disconnected)
    
    var clientId: String { clientIdValue }
    var respTopicPrefix: String { "\(respPrefixBase)\(clientIdValue)/" }
    var inboxTopic: String { "\(inboxPrefixBase)\(employeeId)" }
    
    init(brokerHost: String, brokerPort: Int, useTls: Bool, username: String, password: String, employeeId: String, clientId: String? = nil, serviceId: String? = nil) {
        self.brokerHost = brokerHost
        self.brokerPort = UInt16(min(max(brokerPort, 1), 65535))
        self.useTls = useTls
        self.username = username
        self.password = password
        self.employeeId = employeeId.trimmingCharacters(in: .whitespacesAndNewlines)
        self.clientIdValue = (clientId?.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap { $0.isEmpty ? nil : $0 }
            ?? "\(employeeId)_ios_\(UUID().uuidString.prefix(8))"
        self.serviceIdValue = serviceId?.trimmingCharacters(in: .whitespacesAndNewlines)
        self.topicPrefix = getTopicPrefix(serviceId)
        self.reqPrefix = "\(self.topicPrefix)/msg/req/"
        self.respPrefixBase = "\(self.topicPrefix)/msg/resp/"
        self.inboxPrefixBase = "\(self.topicPrefix)/inbox/"
    }
    
    func setInboxCallback(_ callback: @escaping (String) -> Void) {
        onInboxMessage = callback
    }
    
    fileprivate func setState(_ s: MChatConnectionState) {
        state = s
    }
    
    func connect() {
        userDisconnect = false
        reconnectTask?.cancel()
        reconnectTask = nil
        performConnect()
    }
    
    func disconnect() {
        userDisconnect = true
        reconnectTask?.cancel()
        reconnectTask = nil
        mqtt?.disconnect()
        mqtt = nil
        for (_, cont) in pending {
            cont.resume(returning: MqttResponse(code: 500, message: "Disconnected", data: nil))
        }
        pending.removeAll()
        state = .disconnected
    }
    
    func request(action: String, params: [String: Any]) async -> MqttResponse {
        guard let mqtt = mqtt, mqtt.connState == .connected else {
            return MqttResponse(code: 500, message: "Not connected", data: nil)
        }
        let seqId = "seq_\(UUID().uuidString.prefix(8))_\(Int64(Date().timeIntervalSince1970 * 1000))"
        let topic = "\(reqPrefix)\(clientIdValue)/\(seqId)"
        var body: [String: Any] = ["action": action]
        for (k, v) in params { body[k] = v }
        guard let payloadData = try? JSONSerialization.data(withJSONObject: body),
              let payload = String(data: payloadData, encoding: .utf8) else {
            return MqttResponse(code: 400, message: "Invalid params", data: nil)
        }
        return await withCheckedContinuation { cont in
            pending[seqId] = cont
            mqtt.publish(topic, withString: payload, qos: .qos1)
            Task { @MainActor in
                try? await Task.sleep(nanoseconds: UInt64(requestTimeout * 1_000_000_000))
                if let c = pending.removeValue(forKey: seqId) {
                    c.resume(returning: MqttResponse(code: 504, message: "Request timeout", data: nil))
                }
            }
        }
    }
    
    private func performConnect() {
        state = .connecting
        // brokerUrl for logging purposes
        _ = "\(useTls ? "ssl" : "tcp")://\(brokerHost):\(brokerPort)"
        let mqtt = CocoaMQTT(clientID: clientIdValue, host: brokerHost, port: brokerPort)
        mqtt.username = username
        mqtt.password = password
        mqtt.keepAlive = 60
        mqtt.autoReconnect = false
        mqtt.cleanSession = true
        mqtt.enableSSL = useTls
        mqtt.allowUntrustCACertificate = useTls
        mqtt.delegate = MChatMQTTDelegate(connection: self)
        self.mqtt = mqtt
        
        guard mqtt.connect() else {
            state = .error("Connect failed")
            scheduleReconnect()
            return
        }
    }
    
    fileprivate func didConnect() {
        mqtt?.subscribe("\(respTopicPrefix)+", qos: .qos1)
        mqtt?.subscribe(inboxTopic, qos: .qos1)
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 300_000_000)
            let bindRes = await request(action: "auth.bind", params: ["employee_id": employeeId])
            if bindRes.code != 0 {
                state = .error("auth.bind: \(bindRes.message)")
                return
            }
            state = .connected
            reconnectBackoff = initialReconnect
        }
    }
    
    fileprivate func didDisconnect(_ error: Error?) {
        mqtt = nil
        for (_, cont) in pending {
            cont.resume(returning: MqttResponse(code: 500, message: "Disconnected", data: nil))
        }
        pending.removeAll()
        state = .disconnected
        if !userDisconnect { scheduleReconnect() }
    }
    
    fileprivate func messageArrived(topic: String, payload: String) {
        let cleaned = payload.replacingOccurrences(of: "\u{0}", with: "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleaned.isEmpty else { return }
        if topic.hasPrefix(respTopicPrefix) {
            let seqId = String(topic.dropFirst(respTopicPrefix.count))
            if let cont = pending.removeValue(forKey: seqId) {
                let code = parseCode(cleaned)
                let msg = parseMessage(cleaned)
                let data = parseData(cleaned)
                cont.resume(returning: MqttResponse(code: code, message: msg, data: data))
            }
        } else if topic == inboxTopic {
            onInboxMessage?(cleaned)
        }
    }
    
    private func scheduleReconnect() {
        reconnectTask?.cancel()
        let delay = reconnectBackoff
        reconnectBackoff = min(reconnectBackoff * 2, maxReconnect)
        reconnectTask = Task {
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            guard !userDisconnect else { return }
            await MainActor.run { performConnect() }
        }
    }
    
    private func parseCode(_ payload: String) -> Int {
        guard let range = payload.range(of: "\"code\"") else { return 500 }
        let after = payload[range.upperBound...]
        guard let colon = after.firstIndex(of: ":") else { return 500 }
        let numPart = after[after.index(after: colon)...].prefix(10)
        return Int(String(numPart.filter { $0.isNumber || $0 == "-" })) ?? 500
    }
    
    private func parseMessage(_ payload: String) -> String {
        guard let range = payload.range(of: "\"message\"") else { return "" }
        let after = String(payload[range.upperBound...])
        guard let q1 = after.firstIndex(of: "\""),
              let q2 = after[after.index(after: q1)...].firstIndex(of: "\"") else { return "" }
        return String(after[after.index(after: q1)..<q2])
    }
    
    private func parseData(_ payload: String) -> String? {
        guard let range = payload.range(of: "\"data\"") else { return nil }
        let after = String(payload[range.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
        guard after.first == ":" else { return nil }
        let rest = String(after.dropFirst()).trimmingCharacters(in: .whitespacesAndNewlines)
        if rest.hasPrefix("null") { return nil }
        if rest.hasPrefix("{") || rest.hasPrefix("[") {
            var depth = 0
            for (i, c) in rest.enumerated() {
                if c == "{" || c == "[" { depth += 1 }
                else if c == "}" || c == "]" {
                    depth -= 1
                    if depth == 0 { return String(rest.prefix(i + 1)) }
                }
            }
        }
        return nil
    }
}

/// CocoaMQTT 回调桥接
private class MChatMQTTDelegate: NSObject, CocoaMQTTDelegate {
    private weak var connection: MChatConnection?
    init(connection: MChatConnection) { self.connection = connection }
    
    func mqtt(_ mqtt: CocoaMQTT, didConnectAck ack: CocoaMQTTConnAck) {
        guard ack == .accept else {
            Task { @MainActor in
                connection?.setState(.error("Connect rejected"))
            }
            return
        }
        Task { @MainActor in
            connection?.didConnect()
        }
    }
    
    func mqttDidDisconnect(_ mqtt: CocoaMQTT, withError err: Error?) {
        Task { @MainActor in
            connection?.didDisconnect(err)
        }
    }
    
    func mqtt(_ mqtt: CocoaMQTT, didReceiveMessage message: CocoaMQTTMessage, id: UInt16) {
        let topic = message.topic
        let payload = message.string ?? ""
        Task { @MainActor in
            connection?.messageArrived(topic: topic, payload: payload)
        }
    }
    
    func mqtt(_ mqtt: CocoaMQTT, didSubscribeTopics success: NSDictionary, failed: [String]) {}
    func mqtt(_ mqtt: CocoaMQTT, didUnsubscribeTopics topics: [String]) {}
    func mqtt(_ mqtt: CocoaMQTT, didPublishMessage message: CocoaMQTTMessage, id: UInt16) {}
    func mqtt(_ mqtt: CocoaMQTT, didPublishAck id: UInt16) {}
    func mqttDidPing(_ mqtt: CocoaMQTT) {}
    func mqttDidReceivePong(_ mqtt: CocoaMQTT) {}
}
